torch/csrc/autograd/python_variable.cpp

```diff
@@ -0,0 +1,353 @@
+
+  if (creator == Py_None)
+    creator = NULL;
+
+  THPUtils_assert(!(is_volatile && requires_grad),
+          "Variable can't be volatile and require_grad at the same time!");
+  THPUtils_assert(!creator || THPFunction_Check(creator),
+          "Variable creator has to be a Function object or None, but got %s",
+          THPUtils_typename(creator));
+  THPUtils_assert(THPModule_isTensor(data), "Variable data has to "
+          "be a tensor, but got %s", THPUtils_typename(data));
+
+  return THPVariable_NewWithType(type, data, creator, requires_grad, is_volatile);
+}
+
+typedef PyObject *(*getter)(PyObject *, void *);
+typedef int (*setter)(PyObject *, PyObject *, void *);
+
+PyObject *THPVariable_get_version(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  return PyInt_FromLong(**var.version_counter);
+}
+
+PyObject *THPVariable_get_creator(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  if (!var.creator) {
+    Py_RETURN_NONE;
+  }
+  return functionToPyObject(var.creator);
+}
+
+PyObject * THPVariable_get_data(THPVariable *self)
+{
+  if (!self->data) {
+    auto& var = *self->cdata;
+    PyTypeObject* type = torch::getPyTypeObject(*var.data);
+    self->data = type->tp_alloc(type, 0);
+    if (self->data) {
+      ((torch::THPVoidTensor*)self->data)->cdata =
+          (torch::THVoidTensor *)var.data->retain().cdata();
+    }
+  }
+  Py_INCREF(self->data);
+  return self->data;
+}
+
+int THPVariable_set_data(THPVariable *self, PyObject *data)
+{
+  THPUtils_assertRet(-1, THPModule_isTensor(data), "Variable data has to "
+      "be a tensor, but got %s", THPUtils_typename(data));
+  Py_INCREF(data);
+  Py_XDECREF(self->data);
+  self->data = data;
+  auto& var = *self->cdata;
+  auto tensor = torch::createTensor(data);
+  var.data.swap(tensor);
+  return 0;
+}
+
+PyObject *THPVariable_get_raw_grad(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  if (!var.grad) {
+    Py_RETURN_NONE;
+  }
+  return THPVariable_Wrap(var.grad);
+}
+
+int THPVariable_set_raw_grad(THPVariable *self, PyObject *data)
+{
+  auto& var = *self->cdata;
+  if (data == Py_None) {
+    var.grad.reset();
+    return 0;
+  }
+  THPUtils_assertRet(-1, THPVariable_Check(data),
+      "expected Variable or None (got %s)", THPUtils_typename(data));
+  var.grad = ((THPVariable*)data)->cdata;
+  return 0;
+}
+
+PyObject *THPVariable_get_grad(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  if (!var.grad) {
+#ifdef WITH_CUDA
+    THCPAutoGPU __guard(var.data->getDevice());
+#endif
+    auto grad = var.data->newTensor();
+    grad->resizeAs(*var.data).zero();
+    var.grad = std::make_shared<Variable>(std::move(grad), 0, 1);
+  }
+  return THPVariable_Wrap(var.grad);
+}
+
+PyObject *THPVariable_get_volatile(THPVariable *self)
+{
+  auto& var = *self->cdata;
+  return PyBool_FromLong(var.is_volatile);
+}
+
+int THPVariable_set_volatile(THPVariable *self, PyObject *obj)
+{
+  THPUtils_assertRet(-1, PyBool_Check(obj), "volatile must be a bool");
```

Comment:
Actually can you raise an error if someone sets `volatile=True` on a non-leaf Variable?